relay-server/src/actors/processor/span.rs

```diff
@@ -0,0 +1,418 @@
+            return ItemAction::DropSilently;
+        };
+
+        if let Some(config) = span_metrics_extraction_config {
+            let metrics = extract_metrics(span, config);
+            state.extracted_metrics.project_metrics.extend(metrics);
+        }
+
+        // TODO: dynamic sampling
+
+        if let Err(e) = scrub(&mut annotated_span, &state.project_state.config) {
+            relay_log::error!("failed to scrub span: {e}");
+        }
+
+        // TODO: rate limiting
+
+        // Remove additional fields.
+        process_value(
+            &mut annotated_span,
+            &mut RemoveOtherProcessor,
+            ProcessingState::root(),
+        )
+        .ok();
+
+        // Validate for kafka (TODO: this should be moved to kafka producer)
+        let annotated_span = match validate(annotated_span) {
+            Ok(res) => res,
+            Err(err) => {
+                relay_log::error!("invalid span: {err}");
+                return ItemAction::DropSilently;
+            }
+        };
+
+        // Write back:
+        let mut new_item = Item::new(ItemType::Span);
+        let payload = match annotated_span.to_json() {
+            Ok(payload) => payload,
+            Err(err) => {
+                relay_log::debug!("failed to serialize span: {}", err);
+                return ItemAction::DropSilently;
+            }
+        };
+        new_item.set_payload(ContentType::Json, payload);
+
+        *item = new_item;
+
+        ItemAction::Keep
+    });
+}
+
+/// We do not extract spans with missing fields if those fields are required on the Kafka topic.
+#[cfg(feature = "processing")]
+pub fn validate(mut span: Annotated<Span>) -> Result<Annotated<Span>, anyhow::Error> {
+    let inner = span
+        .value_mut()
+        .as_mut()
+        .ok_or(anyhow::anyhow!("empty span"))?;
+    let Span {
+        ref exclusive_time,
+        ref mut tags,
+        ref mut sentry_tags,
+        ref mut start_timestamp,
+        ref mut timestamp,
+        ref mut span_id,
+        ref mut trace_id,
+        ..
+    } = inner;
+
+    trace_id
+        .value()
+        .ok_or(anyhow::anyhow!("span is missing trace_id"))?;
+    span_id
+        .value()
+        .ok_or(anyhow::anyhow!("span is missing span_id"))?;
+
+    match (start_timestamp.value(), timestamp.value()) {
+        (Some(start), Some(end)) => {
+            if end < start {
+                return Err(anyhow::anyhow!(
+                    "end timestamp is smaller than start timestamp"
+                ));
+            }
+        }
+        (_, None) => {
+            return Err(anyhow::anyhow!("timestamp hard-required for spans"));
+        }
+        (None, _) => {
+            return Err(anyhow::anyhow!("start_timestamp hard-required for spans"));
+        }
+    }
+
+    // `is_segment` is set by `extract_span`.
+    exclusive_time
+        .value()
+        .ok_or(anyhow::anyhow!("missing exclusive_time"))?;
```

Comment:
I changed it so the fallback is to calculate the duration of the span using start and end timestamps. Like this, the validation still works as intended.

Reply:
See other comment:

> This is ultimately a product decision, but I don't think we should assume that a span has zero child spans just because the sentry-specific "exclusive time" attribute is missing. I'd rather drop these spans for now than ingest incorrect data for them.